src/controller/controller.go

```diff
@@ -43,48 +46,102 @@ func NewController(stop chan struct{}) *Controller {
 	eventBroadcaster.StartRecordingToSink(&corev1.EventSinkImpl{Interface: k8sClient.CoreV1().Events("")})
 
 	controller := &Controller{
-		k8sClient:  k8sClient,
-		stopCh:     make(chan struct{}),
-		Stop:       stop,
-		k8sFactory: k8sFactory,
-		updateCh:   make(chan *corev1_api.Pod),
-		recorder:   eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "oom-event-generator"}),
-		podState:   make(map[string]map[string]int32),
-		startTime:  time.Now(),
+		stopCh:         make(chan struct{}),
+		Stop:           stop,
+		k8sFactory:     k8sFactory,
+		eventAddedCh:   make(chan *core.Event),
+		eventUpdatedCh: make(chan *eventUpdateGroup),
+		recorder:       eventBroadcaster.NewRecorder(scheme.Scheme, core.EventSource{Component: "oom-event-generator"}),
+		startTime:      time.Now(),
 	}
 
-	podInformer := informers.SharedInformerFactory(k8sFactory).Core().V1().Pods().Informer()
-	podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
-		UpdateFunc: controller.podUpdated,
-		DeleteFunc: controller.podDeleted,
+	informers.SharedInformerFactory(k8sFactory).Core().V1().Pods().Informer()
+	eventsInformer := informers.SharedInformerFactory(k8sFactory).Core().V1().Events().Informer()
+	eventsInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			controller.eventAddedCh <- obj.(*core.Event)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			controller.eventUpdatedCh <- &eventUpdateGroup{
+				oldEvent: oldObj.(*core.Event),
+				newEvent: newObj.(*core.Event),
+			}
+		},
 	})
-	return controller
-}
-
-func (c *Controller) podUpdated(oldObj, newObj interface{}) {
-	c.updateCh <- newObj.(*corev1_api.Pod)
-}
 
-func (c *Controller) podDeleted(obj interface{}) {
-	c.deletePodState(string(obj.(*corev1_api.Pod).UID))
+	return controller
 }
 
 // Run is the main loop that processes Kubernetes Pod changes
 func (c *Controller) Run() error {
 	c.k8sFactory.Start(c.stopCh)
+	c.k8sFactory.WaitForCacheSync(nil)
```

Comment:
I think it would need to be `c.Stop` as `WaitForCacheSync` is synchronous and the controller doesn't write anything into the `c.stopCh` until after `WaitForCacheSync` completed.

Reply:
Yes, you're right